src/Diff.hs

```diff
@@ -0,0 +1,606 @@
+      , sourcesDiff     :: SourcesDiff
+      , inputExtraNames :: Maybe (Changed (Set Text))
+        -- ^ Will be Nothing, if there is no extra input names
+      , inputsDiff      :: [InputDerivationsDiff]
+      , envDiff         :: Maybe EnvironmentDiff
+        -- ^ Will be Nothing, if environment comparison is skipped
+      }
+
+
+data OutputsDiff = OutputsDiff
+  { extraOutputs :: Maybe (Changed (Map Text (DerivationOutput FilePath Text)))
+    -- ^ Map from derivation name to it's outputs.
+    --   Will be Nothing, if `Data.Map.difference` gives
+    --   empty Maps for both new and old outputs
+  , outputHashDiff :: [OutputDiff]
+    -- ^ Difference of outputs with the same name.
+    --   Will be empty, if all outputs are equal.
+  }
+
+data OutputDiff = OutputDiff
+  { outputName :: Text
+  , hashDifference :: Changed OutputHash
+  }
+
+data SourcesDiff = SourcesDiff
+  { extraSrcNames :: Maybe (Changed (Set Text))
+    -- ^ Will be Nothing, if there is no extra source names
+  , srcFilesDiff :: [SourceFileDiff]
+  }
+
+data SourceFileDiff
+  = OneSourceFileDiff
+      { srcName :: Text
+      , srcContentDiff :: Maybe [Patience.Item Text]
+      -- ^ Will be Nothing, if any of source files not exists
+      }
+  | SomeSourceFileDiff
+      { srcName :: Text
+      , srcFilesDiff :: Changed [FilePath]
+      }
+
+data InputDerivationsDiff
+  = OneDerivationDiff
+      { drvName :: Text
+      , drvDiff :: DerivationDiff
+      }
+  | SomeDerivationsDiff
+      { drvName :: Text
+      , extraPartsDiff :: Changed (Map FilePath (Set Text))
+      }
+
+data EnvVarDiff = EnvVarDiff
+  { envKey :: Text
+  , envValueDiff :: [Patience.Item Text]
+  }
+
+data EnvironmentDiff
+  = EnvironmentsAreEqual
+  | EnvironmentDiff
+      { extraEnvDiff :: Changed (Map Text Text)
+      , envContentDiff :: [EnvVarDiff]
+      }
+
+{-| Extract the name of a derivation (i.e. the part after the hash)
+
+    This is used to guess which derivations are related to one another, even
+    though their hash might differ
+
+    Note that this assumes that the path name is:
+
+    > /nix/store/${32_CHARACTER_HASH}-${NAME}.drv
+
+    Nix technically does not require that the Nix store is actually stored
+    underneath `/nix/store`, but this is the overwhelmingly common use case
+-}
+derivationName :: FilePath -> Text
+derivationName = Text.dropEnd 4 . Text.drop 44 . Text.pack
+
+-- | Group paths by their name
+groupByName :: Map FilePath a -> Map Text (Map FilePath a)
+groupByName m = Data.Map.fromList assocs
+  where
+    toAssoc key = (derivationName key, Data.Map.filterWithKey predicate m)
+      where
+        predicate key' _ = derivationName key == derivationName key'
+
+    assocs = fmap toAssoc (Data.Map.keys m)
+
+{-| Extract the name of a build product
+
+    Similar to `derivationName`, this assumes that the path name is:
+
+    > /nix/store/${32_CHARACTER_HASH}-${NAME}.drv
+-}
+buildProductName :: FilePath -> Text
```

Comment:
What is the point of having both `derivationName` and `buildProductName`? 
Isn't it correct that if `buildProductName a == buildProductName b` then `derivationName a == derivationName b` is also true

Reply:
I don't really know why in one place is used `derivationName` and in the other one - `buildProductName`. This code was in the earlier version of nix-diff, I just didn't change it. As I can see, `buildProductName` is used in diff of sources and this name will be printed in render and the same with `derivationName` and diff of inputs, so I prefer to not change this behavior or to ask code author about it.

Reply:
I don't really insist on changing it since this code existed previously